KTabLists overflow enhancements

AlexVelezLl opened the pull request:
## Description
This PR introduces KListWithOverflow in KTabsList, so KTabsList now shows a "more" button for the overflowed tabs if they dont fit in the available space.

#### Issue addressed
Addresses #623

### Screencasts

https://github.com/learningequality/kolibri/assets/51239030/613e416b-f966-4235-9313-15e620f4e0ee

https://github.com/learningequality/kolibri/assets/51239030/cced15c4-a9cb-4508-8f2c-9a46c223f399

## Changelog

- [#635]
  - **Description:** KListWithOverflow now takes into account the margin of items to calculate available space.
  - **Products impact:** bugfix.
  - **Addresses:** -.
  - **Components:** KListWithOverflow.
  - **Breaking:** no
  - **Impacts a11y:** no
  - **Guidance:** -

- [#635]
  - **Description:** KListWithOverflow now emits an event `updateOverflowItems` when overflowItems are recalculated.
  - **Products impact:** new API.
  - **Addresses:** -.
  - **Components:** KListWithOverflow.
  - **Breaking:** no
  - **Impacts a11y:** no
  - **Guidance:** -.

- [#635]
  - **Description:** Kbutton and KIconButton now emit @focus event.
  - **Products impact:** new API.
  - **Addresses:** -.
  - **Components:** KButton, KIconButton.
  - **Breaking:** no
  - **Impacts a11y:** no
  - **Guidance:** -.

- [#635]
  - **Description:** KTabsList now shows a "more" button for the overflowed tabs if they dont fit in the available space.
  - **Products impact:** updated API.
  - **Addresses:** https://github.com/learningequality/kolibri-design-system/issues/623.
  - **Components:** KTabsList, KTabs.
  - **Breaking:** no
  - **Impacts a11y:** no
  - **Guidance:** -.

- [#635]
  - **Description:** KTabsList now receives a `tabAppearanceOverrides` prop to override styles of tab items.
  - **Products impact:** new API.
  - **Addresses:** .
  - **Components:** KTabsList.
  - **Breaking:** no
  - **Impacts a11y:** no
  - **Guidance:** -.  

[#635]: https://github.com/learningequality/kolibri-design-system/pull/635

## Steps to test

Play with KTabsList and KTabs in the playground and check that it handles overflow properly.

## (optional) Implementation notes

### At a high level, how did you implement this?
<!-- Briefly describe how this works -->

### Does this introduce any tech-debt items?
<!-- List anything that will need to be addressed later -->

## Testing checklist
<!-- Complete the checklist before submitting a PR; delete anything that doesn't apply -->

- [x] Contributor has fully tested the PR manually
- [ ] If there are any front-end changes, before/after screenshots are included
- [ ] Critical and brittle code paths are covered by unit tests
- [ ] The change is described in the changelog section above

## Reviewer guidance
<!-- Delete anything that doesn't apply so your reviewer knows what to check for -->

- [ ] Is the code clean and well-commented?
- [ ] Are there tests for this change?
- [ ] Are all UI components LTR and RTL compliant (if applicable)?
- [ ] _Add other things to check for here_

## After review

- [ ] The changelog item has been pasted to the `CHANGELOG.md`

## Comments
<!-- Any additional notes you'd like to add -->

AlexVelezLl committed:
Use KListWithOverflow inside KTabsList

AlexVelezLl committed:
Partial

AlexVelezLl committed:
Add checks to ktabs

AlexVelezLl committed:
support for tabsWithOverflow

AlexVelezLl committed:
Emit kbutton focus event

AlexVelezLl committed:
Add changelog and fix tests

AlexVelezLl commented:
I am not completely sure if this should be retarget to `develop` since @MisRob suggested to not introduce `KListWithOverflow` in `0.16.x`, and introducing it into `KTabsList` will introduce them in 0.16.x

AlexVelezLl commented on lib/tabs/KTabsList.vue:
```diff
@@ -63,7 +65,22 @@
+      <KIconButton
+        tooltip="More"
+        icon="optionsHorizontal"
+        :style="overflowButtonStyles()"
```
An improvement we can discuss is how can we improve the visibility of the selected tab when it is overflowed. But I think this can be a follow-up issue.

MisRob commented:
> I am not completely sure if this should be retarget to develop since @MisRob suggested to not introduce KListWithOverflow in 0.16.x, and introducing it into KTabsList will introduce them in 0.16.x

Yes, that's right, `develop` would be the right target here, thank you

MisRob commented:
@AlexVelezLl could you please loop in @radinamatic for QA and give her directions for testing to confirm [these acceptance criteria](https://github.com/learningequality/kolibri-design-system/issues/623)?

 
![Screenshot from 2024-05-13 17-29-29](https://github.com/learningequality/kolibri-design-system/assets/13509191/e39b8333-b95a-4736-bac4-77cb93003f6f)

nucleogenesis commented on lib/tabs/KTabsList.vue:
```diff
@@ -63,7 +65,22 @@
-  </div>
+    <template #more="{ overflowItems }">
+      <KIconButton
+        tooltip="More"
```
This should be a prop -- something like `moreButtonTooltip` because we don't have any i18n machinery setup in KDS.

So in Kolibri or Studio, we would need a translation we can pass into that prop like:

`<KTabsList :moreButtonTooltip="coreStrings('moreLabel')" /* ... */>`

That way we'd pass in the translated version of the string to be used as the tooltip label.

nucleogenesis reviewed (changes requested):
One change requested related to i18n stuff. 

I'll make time tomorrow to pull this down and run Kolibri with it to give it a spin.

AlexVelezLl committed:
Add moreTooltip prop to KTabs

AlexVelezLl commented:
Thank you @nucleogenesis! I have added the prop, and updated the [Kolibri PR](https://github.com/learningequality/kolibri/pull/12146/commits/3f3eae69d7a9c2214965d60c5f9e04c285aec5f6) to include it.

MisRob commented:
@AlexVelezLl Thanks a lot, this will be such a valuable update in many places, and it also confirms that `KTabsListWithOverflow` is easy and flexible for use.

I won't manage to do a full review here but I see @nucleogenesis is already doing some code review, and @radinamatic has started a11y testing. As soon as the acceptance criteria I referenced are satisfied, I feel okay about proceeding.

I tested briefly and noticed the following issues. I used Firefox (may be present in Chrome too but I haven't tried yet). Some of them will perhaps be related.

**(1) The responsiveness doesn't work fully and also at some point focus ring strangely appears on its own, even though I didn't use the keyboard here.** Could point to some unexpected troublesome issues in the implementation.

![Peek 2024-05-18 15-44](https://github.com/learningequality/kolibri-design-system/assets/13509191/0dfa5535-0078-482e-85ab-aca59c5de812)

 **(2) As soon as calculation is triggered, something unexpected happens, and all tabs are hidden to the menu instead of gradually one by one** starting with "Groups" which the only one I'd expect to be moved to the context menu here:

![PEEK-2](https://github.com/learningequality/kolibri-design-system/assets/13509191/c6d1c344-ea50-494d-a329-65de93bc9b4b)

(3) **When I click TAB when in the context menu, the focus shifts to the next item and immediately after the menu gets closed by its own.** I am not entirely sure what's expected keyboard navigation for this context menu (arrow keys work well) but pretty certain that we'll need to improve this. @radinamatic will know more about what's expected behavior. In this recording I
- TAB to the context menu icon
- ENTER to open the menu
- then tried to TAB again (causing the issue I'm describing above)

Also note that somehow we have two focus rings in the menu.
 
![peek-3](https://github.com/learningequality/kolibri-design-system/assets/13509191/b8ea8003-5511-4a27-a9a2-4ed1b1ec3d52)

AlexVelezLl commented:
Hi @MisRob! Thanks for your review.

1. This is not a focus ring, that border indicates that the selected tab is hidden in the more button. I followed the implementation that we had in EQM for pointing that the selected tab was hidden. I commented that [here](https://github.com/learningequality/kolibri-design-system/pull/635#discussion_r1595612589) because yes, it can be confusing, and I think the communication that the selected tab is hidden can be improved.
2. Havent tested in the kds docs, and havent catch this. But this happens because currently KListWithOverflow doesnt have support for `inline-block` parent elements, because its available width can't be determined easily. I have tried a couple of options reading the parent's parent's width but it has a lot of edge cases to calculate the available space :c. So for now the use of `display: flex` and setting a `max-height` is needed if we want to achieve something similar to inline-block. I have just updated the docs examples with this, let me know what are your thoughts on this (It's also worth saying that this KListWithOverflow restriction has yet to cause problems where we've used it, and I didnt found any inline-block parents in Kolibri's KTabLists either).
3. This is a known issue we have with KDropdownMenu (https://github.com/learningequality/kolibri-design-system/issues/588), and in previous discussions [in slack](https://learningequality.slack.com/archives/C05FSV4BEJ1/p1713820270887209) it was determined that @tab should return the focus to the next focusable element. And the two focus rings is because one is the focus ring, and the other is the border describing that the selected tab is hidden.

AlexVelezLl committed:
Make docsShow display block

MisRob commented:
Thanks for looking into it @AlexVelezLl.

(1) Ah yes, I didn't realize the connection, and agree it's fine to do it as a follow-up issue
(2) I see. I think it's fine to operate within some limits as long as it's communicated. For calculations that need width, it is very much understandable that parent needs to have the width information. Generally speaking, I haven't found a reasonable way to do width calculations without that either and agree that it's better to ensure the component gets rendered in contexts with width available. One example would be ratio calculations on `KImg` where I made a note about it in the documentation _"Note that the ratio styles calculations need to have the width information, therefore it needs to be available in some way. For example, you can set the width directly on KImg. Alternatively, you could ensure that KImg's parent element has width by setting it explicitly or by using a block element."_ Could you add a similar guidance to
- `KListWithOverflow` docs - detailed guidance
- `KTabsList` brief note and reference to `KListWithOverflow` docs
- `KTabs` brief note and reference to `KListWithOverflow` docs

(3) I see, all good then. Also perhaps the discussion you have with @radinamatic in the related Kolibri PR will help to resolve this.

MisRob commented:
Now I also noticed that icons (and more generally contents of the `tabs` slot) don't get transferred to the menu so we may need to look into that at some point. I think that's a further improvement that would be fine to have a follow-up issue for.

![Screenshot from 2024-05-21 11-09-12](https://github.com/learningequality/kolibri-design-system/assets/13509191/b9ec3b7d-21a0-4f54-8287-0c38cd45db1a)

AlexVelezLl committed:
Navigate through list and dropdown button

AlexVelezLl committed:
Support left/right arrow navigations in kdrokdropdownmenu

AlexVelezLl committed:
Lint files

AlexVelezLl committed:
Lint files

AlexVelezLl committed:
Handle isMoreButtonFocused

AlexVelezLl committed:
isSelected Support in KdropdownMenu options

AlexVelezLl committed:
Fix tests

AlexVelezLl committed:
Test tab role

AlexVelezLl committed:
Rol + roledescription

AlexVelezLl committed:
Fix aria-selected

AlexVelezLl commented:
Closing since another approach will be implemented to handle overflow in tabs. See https://github.com/learningequality/kolibri/pull/12146#issuecomment-2127951322
